### PR TITLE
feat(core): use the CSS parser to shim Shadow DOM CSS code

### PR DESCRIPTION
**Pull request by matsko**
Closes #9154

---

@vicb there is a mix of variables/classes with `AST` and `Ast`. Once this PR goes in then I'll make a new PR that converts everything to `Ast` ... Same thing with `Css`.

### CODE REVIEW

File: modules/@angular/compiler/src/css/lexer.ts [hunk 1]
```diff
@@ -178,21 +180,30 @@ export class CssScanner {
 
   consume(type: CssTokenType, value: string = null): LexedCssResult {
     var mode = this._currentMode;
-    this.setMode(CssLexerMode.ALL);
+
+    var tempMode = CssLexerMode.ALL;
+    if (_trackWhitespace(mode)) {
+      tempMode = CssLexerMode.ALL_TRACK_WS;
```
Comment (vicb) [inline modules/@angular/compiler/src/css/lexer.ts]: simplify: `if() { this.setMode }`

File: modules/@angular/compiler/src/css/parser.ts [hunk 2]
```diff
@@ -68,22 +91,28 @@ function characterContainsDelimiter(code: number, delimiters: number): boolean {
   return bitWiseAnd([getDelimFromCharacter(code), delimiters]) > 0;
 }
 
-export class CssAST {
-  visit(visitor: CssASTVisitor, context?: any): void {}
+export abstract class CssAST {
+  constructor(public start: number, public end: number) {}
+  visit(visitor: CssASTVisitor, context?: any): any { return null; }
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: Do you want a default impl ?

File: modules/@angular/facade/src/lang.dart [hunk 3]
```diff
@@ -397,6 +397,11 @@ num bitWiseAnd(List values) {
   return val as num;
 }
 
+num bitWiseNot(num value) {
+  int val = (value as int);
```
Comment (vicb) [inline modules/@angular/facade/src/lang.dart]: do the tests pass :)

File: modules/@angular/compiler/src/css/parser.ts [hunk 4]
```diff
@@ -28,6 +30,24 @@ const COLON_DELIM = 16;
 const SEMICOLON_DELIM = 32;
 const NEWLINE_DELIM = 64;
 const RPAREN_DELIM = 128;
+const LPAREN_DELIM = 256;
+const SPACE_DELIM = 512;
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: `xxx_FLAG` or `IS_xxx` ?

File: modules/@angular/compiler/src/css/parser.ts [hunk 5]
```diff
@@ -28,6 +30,24 @@ const COLON_DELIM = 16;
...
+
+function _pseudoSelectorSupportsInnerSelectors(name: string) {
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: return type

File: modules/@angular/compiler/src/css/parser.ts [hunk 6]
```diff
@@ -150,14 +179,16 @@ export class CssParser {
   }
 
   /** @internal */
-  _parseStyleSheet(delimiters: any /** TODO #9100 */): CssStyleSheetAST {
+  _parseStyleSheet(delimiters: number): CssStyleSheetAST {
+    var start = this._getScannerIndex();
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: const ?

File: modules/@angular/compiler/src/css/parser.ts [hunk 7]
```diff
@@ -170,82 +201,116 @@ export class CssParser {
...
         var name = tokens[0];
-        return new CssKeyframeRuleAST(name, this._parseKeyframeBlock(delimiters));
+        end = this._scanner.index - 1;
+        return new CssKeyframeRuleAST(start, end, name, this._parseKeyframeBlock(delimiters));
 
       case BlockType.MediaQuery:
         this._scanner.setMode(CssLexerMode.MEDIA_QUERY);
         var tokens = this._collectUntilDelim(bitWiseOr([delimiters, RBRACE_DELIM, LBRACE_DELIM]));
-        return new CssMediaQueryRuleAST(tokens, this._parseBlock(delimiters));
+        end = this._scanner.index - 1;
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: same as `this._getSCannerIndex();` ? use either consistently

File: modules/@angular/compiler/src/css/parser.ts [hunk 8]
```diff
@@ -608,91 +843,172 @@ export class CssParser {
 }
 
 export class CssStyleValueAST extends CssAST {
-  constructor(public tokens: CssToken[], public strValue: string) { super(); }
-  visit(visitor: CssASTVisitor, context?: any) { visitor.visitCssValue(this); }
+  constructor(start: number, end: number, public tokens: CssToken[], public strValue: string) {
```
Comment (tbosch) [inline modules/@angular/compiler/src/css/parser.ts]: Use the `ParseSourceSpan`, a utility class that we already use for template compilation. See `@angular/compiler/util.ts`

Comment (tbosch) [inline modules/@angular/compiler/src/css/parser.ts]: (instead of the two fields `start/end` have just one field `sourceSpan: ParseSourceSpan`)

Comment (tbosch) [inline modules/@angular/compiler/src/css/parser.ts]: This e.g. also contains the link to the source file, ...

File: modules/@angular/compiler/src/css/parser.ts [hunk 9]
```diff
@@ -324,111 +400,254 @@ export class CssParser {
...
+    var lastOperatorToken: CssToken = null;
+    while (operator == null && !characterContainsDelimiter(this._scanner.peek, delimiters) &&
+           isSelectorOperatorCharacter(this._scanner.peek)) {
+      var token = this._scan();
+      var tokenOperator = token.strValue;
+      operatorScanCount++;
+      lastOperatorToken = token;
+      if (tokenOperator != SPACE_OPERATOR) {
+        switch (tokenOperator) {
+          case '/':
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: don't you have const defined ?

File: modules/@angular/compiler/src/css/parser.ts [hunk 10]
```diff
@@ -608,91 +843,172 @@ export class CssParser {
...
+      public query: CssAtRulePredicateAST, block: CssBlockAST) {
+    super(start, end, type, block);
+    var firstCssToken: CssToken = query.tokens[0];
     this.name = new CssToken(
         firstCssToken.index, firstCssToken.column, firstCssToken.line, CssTokenType.Identifier,
         this.strValue);
   }
-  visit(visitor: CssASTVisitor, context?: any) { visitor.visitCssBlock(this.block, context); }
+  visit(visitor: CssASTVisitor, context?: any): any {
+    return visitor.visitCssBlock(this.block, context);
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: does not match the class name

File: modules/@angular/compiler/src/css/parser.ts [hunk 11]
```diff
@@ -608,91 +843,172 @@ export class CssParser {
...
 }
 
 export class CssInlineRuleAST extends CssRuleAST {
-  constructor(public type: BlockType, public value: CssStyleValueAST) { super(); }
-  visit(visitor: CssASTVisitor, context?: any) { visitor.visitInlineCssRule(this, context); }
+  constructor(start: number, end: number, public type: BlockType, public value: CssStyleValueAST) {
+    super(start, end);
+  }
+  visit(visitor: CssASTVisitor, context?: any): any {
+    return visitor.visitInlineCssRule(this, context);
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: mismatch

File: modules/@angular/compiler/src/css/parser.ts [hunk 12]
```diff
@@ -608,91 +843,172 @@ export class CssParser {
...
+}
+
+export class CssSelectorAST extends CssSelectorPartAST {
+  public strValue: string;
+  constructor(start: number, end: number, public selectorParts: CssSimpleSelectorAST[]) {
+    super(start, end);
+    this.strValue = selectorParts.map(part => part.strValue).join('');
+  }
+  visit(visitor: CssASTVisitor, context?: any): any {
+    return visitor.visitCssSelector(this, context);
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: mismattch

File: modules/@angular/compiler/src/css/parser.ts [hunk 13]
```diff
@@ -708,7 +1024,20 @@ export class CssParseError extends ParseError {
   constructor(span: ParseSourceSpan, message: string) { super(span, message); }
 }
 
+export class CssUnknownRuleAST extends CssRuleAST {
+  constructor(start: number, end: number, public ruleName: string, public tokens: CssToken[]) {
+    super(start, end);
+  }
+  visit(visitor: CssASTVisitor, context?: any): any {
+    return visitor.visitUnknownRule(this, context);
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: mismatch

File: modules/@angular/compiler/src/css/parser.ts [hunk 14]
```diff
@@ -708,7 +1024,20 @@ export class CssParseError extends ParseError {
...
+}
+
 export class CssUnknownTokenListAST extends CssRuleAST {
-  constructor(public name: any /** TODO #9100 */, public tokens: CssToken[]) { super(); }
-  visit(visitor: CssASTVisitor, context?: any) { visitor.visitUnkownRule(this, context); }
+  constructor(start: number, end: number, public name: string, public tokens: CssToken[]) {
+    super(start, end);
+  }
+  visit(visitor: CssASTVisitor, context?: any): any {
+    return visitor.visitUnknownTokenList(this, context);
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: mismatch

File: modules/@angular/compiler/src/shadow_css.ts [hunk 15]
```diff
@@ -137,382 +141,775 @@ export class ShadowCss {
...
-  shimCssText(cssText: string, selector: string, hostSelector: string = ''): string {
+  shimCssText(cssText: string, scopeSelector: string, hostSelector: string = ''): string {
     cssText = stripComments(cssText);
-    cssText = this._insertDirectives(cssText);
-    return this._scopeCssText(cssText, selector, hostSelector);
+
+    var cssFileName = 'some-fake-file-name.css';  // TODO (matsko): get the name in here
+    var output = _parseCssCode(cssFileName, cssText);
+
+    // var errors = output.errors;
```
Comment (vicb) [inline modules/@angular/compiler/src/shadow_css.ts]: debug

File: modules/@angular/compiler/src/css/parser.ts (hunk 4 above)
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: SEMICOLON_DELIM_FLAG

File: modules/@angular/compiler/src/css/parser.ts [hunk 16]
```diff
@@ -59,6 +79,9 @@ function getDelimFromCharacter(code: number): number {
       return LBRACE_DELIM;
     case $RPAREN:
       return RPAREN_DELIM;
+    case $SPACE:
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: add Flag to fn name

File: modules/@angular/compiler/src/css/parser.ts (hunk 2 above)
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: abstract

File: modules/@angular/compiler/src/css/parser.ts [hunk 17]
```diff
@@ -324,111 +400,254 @@ export class CssParser {
...
+          // to the CSS specification
+          innerDelims = bitWiseOr([innerDelims, COMMA_DELIM]);
+        }
+
+        // :host(a, b, c) {
+        this._parseSelectors(innerDelims).forEach((selector, index) => {
+          if (index > 0) {
+            strValue += ', ';
+          }
+          strValue += selector.strValue;
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: substring

File: modules/@angular/compiler/src/css/parser.ts [hunk 18]
```diff
@@ -324,111 +400,254 @@ export class CssParser {
...
+      }
+
+      var closeParenToken = this._consume(CssTokenType.Character, ')');
+      tokens.push(closeParenToken);
+      strValue += closeParenToken.strValue;
+    }
+
+    var end = this._getScannerIndex() - 1;
+    return new CssPseudoSelectorAST(
+        start, end, strValue, pseudoSelectorName, isPseudoElement, tokens, innerSelectors);
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: -isPseudoElement

File: modules/@angular/compiler/src/css/parser.ts [hunk 19]
```diff
@@ -324,111 +400,254 @@ export class CssParser {
...
+  }
+
+  /** @internal */
+  _parseSimpleSelector(delimiters: number): CssSimpleSelectorAST {
+    var start = this._getScannerIndex();
+
+    delimiters = bitWiseOr([delimiters, COMMA_DELIM]);
+
     this._scanner.setMode(CssLexerMode.SELECTOR);
+    var strValue = '';
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: substring

File: modules/@angular/compiler/src/css/parser.ts [hunk 20]
```diff
@@ -324,111 +400,254 @@ export class CssParser {
...
-          }
+          // we set the mode after the scan because attribute mode does not
+          // allow attribute [] values. And this also will catch any errors
+          // if an extra "[" is used inside.
+          strValue += '[';
+          selectorCssTokens.push(this._scan());
+          this._scanner.setMode(CssLexerMode.ATTRIBUTE_SELECTOR);
           break;
 
         case $RBRACKET:
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: check for ] at the beginning

File: modules/@angular/compiler/src/css/parser.ts [hunk 21]
```diff
@@ -502,19 +727,26 @@ export class CssParser {
...
     this._scanner.consumeEmptyStatements();
 
-    return new CssBlockAST(results);
+    var end = this._getScannerIndex() - 1;
+    return new CssBlockAST(start, end, results);
   }
 
   /** @internal */
-  _parseStyleBlock(delimiters: number): CssBlockAST {
+  _parseStyleBlock(delimiters: number): CssStylesBlockAST {
```
Comment (vicb) [inline modules/@angular/compiler/src/css/parser.ts]: add comment to AST

File: modules/@angular/compiler/test/css/parser_spec.ts [hunk 22]
```diff
@@ -458,8 +474,12 @@ export function main() {
...
       expect(rule1.selectors.length).toEqual(1);
 
-      var selector = rule1.selectors[0];
-      assertTokens(selector.tokens, ['div', ':', 'not', '(', '.', 'ignore-this-div', ')']);
+      var simpleSelector = rule1.selectors[0].selectorParts[0];
+      assertTokens(simpleSelector.tokens, ['div']);
+
+      var pseudoSelector = simpleSelector.pseudoSelectors[0];
+      expect(pseudoSelector.name).toEqual('not');
+      assertTokens(pseudoSelector.tokens, ['.', 'ignore-this-div']);
```
Comment (vicb) [inline modules/@angular/compiler/test/css/parser_spec.ts]: Add a test to test inner selectors (:host-context)

File: modules/@angular/compiler/test/css/visitor_spec.ts [hunk 23]
```diff
@@ -134,15 +161,24 @@ export function main() {
 
       var rule1 = <CssSelectorRuleAST>captures[0][0];
```
Comment (vicb) [inline modules/@angular/compiler/test/css/visitor_spec.ts]: helper

File: modules/@angular/compiler/test/css/visitor_spec.ts [hunk 24]
```diff
@@ -219,5 +265,47 @@ export function main() {
...
+
+      expect(captures.length).toEqual(1);
+
+      var rule = <CssUnknownRuleAST>captures[0][0];
+      expect(rule.ruleName).toEqual('@someUnknownRule');
+
+      _assertTokens(rule.tokens, ['param', '{', 'one', 'two', 'three', '}']);
+    });
+
+    it('should collect an invalid list of tokens before a valid selector', () => {
```
Comment (vicb) [inline modules/@angular/compiler/test/css/visitor_spec.ts]: group unknown in a describe

File: modules/@angular/facade/src/lang.ts [hunk 25]
```diff
@@ -497,6 +497,10 @@ export function bitWiseAnd(values: number[]): number {
   return values.reduce((a, b) => { return a & b; });
 }
 
+export function bitWiseNot(value: number): number {
```
Comment (vicb) [inline modules/@angular/facade/src/lang.ts]: use inline (after PR)

File: modules/@angular/compiler/src/shadow_css.ts [hunk 26]
```diff
@@ -137,382 +141,775 @@ export class ShadowCss {
...
-                   })
-                   .join(sep);
+  return new CssSelectorTransformer(ast);
+}
+
+function _parseCssCode(cssFileName: string, cssText: string): ParsedCssResult {
+  var lexer = new CssLexer();
+  var scanner = lexer.scan(cssText);
+  var parser = new CssParser(scanner, cssFileName);
+  return parser.parse();
```
Comment (vicb) [inline modules/@angular/compiler/src/shadow_css.ts]:
```
const lexer = new CssLexer();
const parser = CssParser(lexer);
parser.parse(file);
```

File: modules/@angular/compiler/src/shadow_css.ts [hunk 27]
```diff
@@ -137,382 +141,775 @@ export class ShadowCss {
   * be included in the document via WebComponents.ShadowCSS.addCssToDocument(css).
   *
   * When strictStyling is true:
-  * - selector is the attribute added to all elements inside the host,
+  * - scopeSelector is the attribute added to all elements inside the host,
   * - hostSelector is the attribute added to the host itself.
   */
-  shimCssText(cssText: string, selector: string, hostSelector: string = ''): string {
+  shimCssText(cssText: string, scopeSelector: string, hostSelector: string = ''): string {
```
Comment (vicb) [inline modules/@angular/compiler/src/shadow_css.ts]: `shimCss(file, selectors)` parser is injected

keep `shimCssText`, deprecated

File: modules/@angular/compiler/src/shadow_css.ts [hunk 28]
```diff
@@ -137,382 +141,775 @@ export class ShadowCss {
...
+
+    // var errors = output.errors;
+    // if (errors.length > 0) {
+    // throw new BaseException(errors.map((error: CssParseError) => error.msg).join(', '));
+    //}
+
+    var context =
+        new _CssShimSourceContext(cssText, scopeSelector, hostSelector, this.strictStyling);
+
+    var visitor = new _ShadowShimVisitor(cssText, output.ast);
```
Comment (vicb) [inline modules/@angular/compiler/src/shadow_css.ts]: `new _ShadowShimVisitor(output);`

**Pull request closed without merging**